### DISCUSSION TITLE
Vue + Vite + TS - invoking serverless functions locally returns `Error: Cannot use import statement outside a module`

### DISCUSSION THREAD

**Discussion by cohlar** [vercel/vercel#10317] [Help] [answered]
### Question

Hello,

I have set up a Vue project with Vite + TypeScript by running `npm create vue@latest` ([in accordance with vue documentation](https://vuejs.org/guide/quick-start.html#creating-a-vue-application)), and have followed the Vercel documention to [create a Serverless Function](https://vercel.com/docs/concepts/functions/serverless-functions/quickstart) - it works great once deployed ([make a GET call here](https://vercel-vite-vue-seven.vercel.app/api/serverless-example?query=123)).

However I cannot get it to work locally with `vercel dev`, and was hoping I could get some help - [here is the reproduction repo](https://github.com/cohlar/vercel-vite-vue/tree/no-inngest).

Thank you for your help!

## My problem
When trying to run it locally with `vercel dev`, the function invocation fails with error message `Error: Cannot use import statement outside a module` - see screenshots below.

![image](https://github.com/vercel/vercel/assets/57757008/645d4d60-cad7-4073-9011-f6ac66f78643)
![image](https://github.com/vercel/vercel/assets/57757008/9eb8b5d9-e012-4b4d-aa43-18e566227a31)

## My (disappointing) workaround
One workaround I found is to add `"type": "module"` to `package.json` - this makes my simple example work locally, but:
1. Once deployed, the function fails with error message `ERROR	ReferenceError: exports is not defined in ES module scope` ([make a GET call here](https://vercel-vite-vue-dev-brandfolio.vercel.app/api/serverless-example?query=123), and see vercel log screenshot below)
2. ES modules introduce a few other constraints (such as importing with file extension) which I would like to avoid on my larger codebase

![image](https://github.com/vercel/vercel/assets/57757008/57b6b72c-1249-4614-9bb4-ae217031323c)

## Last Comments
I have a feeling this just needs the right TS configuration, and have tried quite a few tweaks to `tsconfig.json`, but so far couldn't make it work.

On a side note, I feel like such template would be nice to have [here](https://vercel.com/docs/frameworks/vite) - I have seen the [vite](https://github.com/vercel/vercel/tree/main/examples/vite) and the [vue](https://github.com/vercel/vercel/tree/main/examples/vue) templates, however none of these include any serverless function, and none uses TypeScript.

Also, I have seen a few other posts about issues with vite and serverless functions (e.g. [this one](https://github.com/vercel/vercel/discussions/7139)), however this doesn't seem related to me, so I created a new discussion.

**Comment by cohlar** [comment, marked as answer]
I just realized that [this section of the documentation](https://vercel.com/docs/functions/serverless-functions/runtimes/node-js#using-typescript-with-the-node.js-runtime) says:

> You can also use a `tsconfig.json` file at the root of your project to configure the TypeScript compiler. Most options are supported aside from ["Path Mappings"](https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping) and ["Project References"](https://www.typescriptlang.org/docs/handbook/project-references.html).

I used a project reference for the TS configuration of the API section which is why it didn't work.
If you just add `"type": "module"` to `package.json` and a `tsconfig.json` in the `./api` directory, it all works 🥳 

Here is a [**working** reproduction repo](https://github.com/cohlar/vercel-vite-vue/tree/no-inngest).

**Reply by TechQuery** [reply]
Thanks for your help! My [One-click Serverless][1] template will make all these things easier, welcome to use & feedback~

[1]: https://github.com/idea2app/Vercel-Node-serverless
